db/hints/sync_point_service.cc

```diff
@@ -0,0 +1,431 @@
+    // This abort_source will be triggered if either of those situations happen.
+    abort_source combined_as;
+    auto trigger_on_operation_abort_sub = as.subscribe([&] () noexcept {
+        if (!combined_as.abort_requested()) {
+            combined_as.request_abort();
+        }
+    });
+
+    co_await parallel_for_each(source_nodes, [this, &combined_as_ = combined_as, id_ = id, &target_nodes_ = target_nodes] (gms::inet_address follower) -> future<> {
+        auto& ms = _messaging;
+        auto& combined_as = combined_as_;
+        const auto id = id_;
+        auto& target_nodes = target_nodes_;
+
+        try {
+            // Create a hint sync point
+            sync_point_create_request request;
+            request.sync_point_id = id;
+            request.target_endpoints = target_nodes;
+            co_await ms.send_hint_sync_point_create(netw::msg_addr{ follower, 0 }, clock_type::time_point::max(), std::move(request));
+
+            // Wait in a loop until the sync point resolves
+            while (true) {
+                sync_point_check_request request;
+                request.sync_point_id = id;
+                auto resp = co_await ms.send_hint_sync_point_check(netw::msg_addr{ follower, 0 }, clock_type::time_point::max(), std::move(request));
+
+                if (resp.expired) {
+                    // Hints were successfully replayed - return
+                    break;
+                }
+
+                // Hints are still being replayed on the follower
+                // Wait a little before checking again
+                // If abort_source was engaged, this will throw - that's what we want
+                co_await sleep_abortable(follower_status_check_interval, combined_as);
+            }
+        } catch (...) {
+            // TODO: Print error?
+            if (!combined_as.abort_requested()) {
+                combined_as.request_abort();
+            }
+            throw;
+        }
+    });
+}
+
+future<> sync_point_service::wait_for_local_hints(
+        abort_source& as,
+        sync_point_id id,
+        std::vector<gms::inet_address> target_nodes) {
+    
+    assert(this_shard_id() == 0);
+
+    // If an exception occurs on one shard, we are no longer interested
+    // in waiting for other shards, so we need a sharded abort source.
+    // Gate is needed because abort_all_shards spawns discarded futures
+    // which needs to be waited on.
+    struct abort_source_and_gate {
+        seastar::gate gate;
+        abort_source as;
+    };
+    sharded<abort_source_and_gate> sharded_abort_source;
+    co_await sharded_abort_source.start();
+
+    auto abort_all_shards = [&sharded_abort_source] {
+        auto& local = sharded_abort_source.local();
+        if (!local.gate.try_enter()) {
+            return;
+        }
+
+        // Waited on by the shard-local gate
+        (void)sharded_abort_source.invoke_on(0, [&] (auto& local0) {
```

Comment:
This comment applies to this whole series: implicit capturing by reference is dangerous, especially for continuations - let's spend a while and explicitly state what's actually captured. In most places it's probably enough to just capture `this`, or even nothing

Reply:
Done, I looked for all `[&]`, `[&,` and `&]`'s and changed them to explicit captures.